src/engine/mod.rs

```diff
@@ -0,0 +1,320 @@
+use self::context::Context;
+use self::equal::RelativeEq;
+use crate::{
+    builtin::{
+        filter::{self, FilterFn},
+        function::{self, FunctionFn},
+    },
+    error::{bail, Result},
+    parser::ast::*,
+    utils::validate_f64,
+};
+use serde_json::{Number, Value};
+use std::collections::HashMap;
+
+pub mod context;
+mod equal;
+
+pub struct Engine {
+    functions: HashMap<String, FunctionFn>,
+    filters: HashMap<String, FilterFn>,
+}
+
+impl Default for Engine {
+    fn default() -> Engine {
+        let mut filters = HashMap::new();
+        filters.insert("upper".into(), Box::new(filter::upper) as FilterFn);
```

Comment:
Just thinking out loud here, no need to do anything :
Would it be possible (and would it make sense?) to do the registration of filters and other functions during compilation time, using e.g. macro ?

Reply:
For what purpose? Once #8 will be implemented this will be removed and public registration functions will be used.